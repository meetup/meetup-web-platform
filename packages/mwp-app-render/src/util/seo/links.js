import React from 'react';
import locales from 'mwp-config/locales';

/**
 * Generates array of React.element's of <link />'s containing canonical + locale urls for the path provided
 * @param  {String} baseUrl base url of the page (protocol + hostname)
 * @param  {String} localeCode   locale of user
 * @param  {String} forcedLocaleCode   locale of group
 * @param  {String} route route currently being viewed
 * @return {String}  composed canonical url
 */
export const generateCanonicalUrl = (
	baseUrl,
	localeCode,
	route,
	forcedLocaleCode
) => {
	const newLocaleCode = forcedLocaleCode || localeCode;
	if (newLocaleCode === 'en-US') {
		return `${baseUrl}${route}`;
	}
	return `${baseUrl}/${newLocaleCode}${route}`;
};

/**
 * Generates array of React.element's of <link />'s containing
 * canonical + locale urls for the path provided
 * @param  {String} baseUrl base url of the page (protocol + hostname)
 * @param  {String} localeCode   locale of user
 * @param  {String} forcedLocaleCode  locale of group
 * @param {Boolean} isGenerateAlternateLinks Defines if we need to generate alternate link tags for the page
 * @param  {String} route   redux'd route to the current page
 * @return {Array}  array of React.element's
 */
export const generateCanonicalUrlLinkTags = (
	baseUrl,
	localeCode,
	route,
	forcedLocaleCode = '',
	isGenerateAlternateLinks
) => {
	let result = [
		<link
			rel="canonical"
			href={generateCanonicalUrl(baseUrl, localeCode, route, forcedLocaleCode)}
			key="canonical"
		/>,
	];

	if (forcedLocaleCode) {
		return result;
	}

	const localeLinks = locales.reduce((acc, locale) => {
		const locationDependentTag = (
			<link
				rel="alternate"
				hrefLang={locale == 'en-US' ? 'en' : locale}
				href={generateCanonicalUrl(baseUrl, locale, route)}
				key={locale}
			/>
		);

		// Google recommends adding location-independent tags for each hreflang tag that includes a location
		// We skip 'es' and 'es-ES' as they are already both supported in locales array
		// and 'en-US' \ 'en-AU' as 'en' hreflang is rendered by default
		if (
			locale !== 'en-US' &&
			locale !== 'en-AU' &&
			locale !== 'es' &&
			locale !== 'es-ES'
		) {
			const localeParts = locale.split('-');
			if (localeParts.length === 2) {
				const locationIndependentTag = (
					<link
						rel="alternate"
						hrefLang={localeParts[0]}
						href={generateCanonicalUrl(baseUrl, locale, route)}
						key={localeParts[0]}
					/>
				);

				return [...acc, locationDependentTag, locationIndependentTag];
			}
		}
		return [...acc, locationDependentTag];
	}, []);

	const alternateLinks = isGenerateAlternateLinks
		? [
				...localeLinks,
				<link
					rel="alternate"
					hrefLang="x-default"
					href={`${baseUrl}${route}`}
					key="default"
				/>,
		  ]
		: '';

	result = [...result, ...alternateLinks];

	return result;
};
