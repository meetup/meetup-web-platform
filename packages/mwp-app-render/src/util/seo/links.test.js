import { generateCanonicalUrlLinkTags } from './links';

describe('generateCanonicalUrlLinkTags', () => {
	const MOCK_BASE_URL = 'http://www.mock-base-url.com';
	const MOCK_ROUTE = '/mockroute';
	const MOCK_LOCALE_CODE = 'fr-FR';
	const MOCK_IS_GENERATE_ALTERNATE_LINKS = true;

	it('should generate locale <link />s for all locales', () => {
		const canonicalUrlMetaTags = generateCanonicalUrlLinkTags(
			MOCK_BASE_URL,
			'en-US',
			MOCK_ROUTE,
			MOCK_IS_GENERATE_ALTERNATE_LINKS
		);
		expect(canonicalUrlMetaTags).toMatchSnapshot();
	});

	it('should generate a canonical link tag with locale if *not* en-US', () => {
		const canonicalUrlMetaTags = generateCanonicalUrlLinkTags(
			MOCK_BASE_URL,
			MOCK_LOCALE_CODE,
			MOCK_ROUTE,
			MOCK_IS_GENERATE_ALTERNATE_LINKS
		);
		expect(
			canonicalUrlMetaTags.filter(el => el.props.rel === 'canonical')
		).toMatchSnapshot();
	});

	it('should generate a canonical link tag without locale if en-US', () => {
		const canonicalUrlMetaTags = generateCanonicalUrlLinkTags(
			MOCK_BASE_URL,
			'en-US',
			MOCK_ROUTE,
			MOCK_IS_GENERATE_ALTERNATE_LINKS
		);
		expect(
			canonicalUrlMetaTags.filter(el => el.props.rel === 'canonical')
		).toMatchSnapshot();
	});

	it('should generate a x-default url that is baseUrl + route', () => {
		const canonicalUrlMetaTags = generateCanonicalUrlLinkTags(
			MOCK_BASE_URL,
			MOCK_LOCALE_CODE,
			MOCK_ROUTE,
			MOCK_IS_GENERATE_ALTERNATE_LINKS
		);
		expect(
			canonicalUrlMetaTags.filter(el => el.props.hrefLang === 'x-default')
		).toMatchSnapshot();
	});
});
