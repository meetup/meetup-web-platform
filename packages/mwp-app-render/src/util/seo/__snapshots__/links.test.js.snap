// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCanonicalUrlLinkTags should generate a canonical link tag with locale if *not* en-US 1`] = `
Array [
  <link
    href="http://www.mock-base-url.com/fr-FR/mockroute"
    rel="canonical"
/>,
]
`;

exports[`generateCanonicalUrlLinkTags should generate a canonical link tag without locale if en-US 1`] = `
Array [
  <link
    href="http://www.mock-base-url.com/mockroute"
    rel="canonical"
/>,
]
`;

exports[`generateCanonicalUrlLinkTags should generate a x-default url that is baseUrl + route 1`] = `
Array [
  <link
    href="http://www.mock-base-url.com/mockroute"
    hreflang="x-default"
    rel="alternate"
/>,
]
`;

exports[`generateCanonicalUrlLinkTags should generate locale <link />s for all locales 1`] = `
Array [
  <link
    href="http://www.mock-base-url.com/mockroute"
    hreflang="en"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/de-DE/mockroute"
    hreflang="de-DE"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/en-AU/mockroute"
    hreflang="en-AU"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/es/mockroute"
    hreflang="es"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/es-ES/mockroute"
    hreflang="es-ES"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/fr-FR/mockroute"
    hreflang="fr-FR"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/it-IT/mockroute"
    hreflang="it-IT"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/ja-JP/mockroute"
    hreflang="ja-JP"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/ko-KR/mockroute"
    hreflang="ko-KR"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/nl-NL/mockroute"
    hreflang="nl-NL"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/pt-BR/mockroute"
    hreflang="pt-BR"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/pl-PL/mockroute"
    hreflang="pl-PL"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/ru-RU/mockroute"
    hreflang="ru-RU"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/th-TH/mockroute"
    hreflang="th-TH"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/tr-TR/mockroute"
    hreflang="tr-TR"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/mockroute"
    hreflang="x-default"
    rel="alternate"
/>,
  <link
    href="http://www.mock-base-url.com/mockroute"
    rel="canonical"
/>,
]
`;
